File: backend/src/main/java/com/radar/prospection/claude/ClaudeService.java
package com.radar.prospection.claude;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.radar.prospection.config.RadarProperties;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class ClaudeService {

    private final RadarProperties properties;
    private final WebClient.Builder webClientBuilder;
    private final ObjectMapper objectMapper;

    private WebClient buildClient() {
        return webClientBuilder
                .baseUrl(properties.getClaude().getBaseUrl())
                .defaultHeader("x-api-key", properties.getClaude().getApiKey())
                .defaultHeader("anthropic-version", "2023-06-01")
                .defaultHeader("content-type", "application/json")
                .build();
    }

    public String complete(String systemPrompt, String userMessage) {
        try {
            Map<String, Object> body = Map.of(
                "model", properties.getClaude().getModel(),
                "max_tokens", 1024,
                "system", systemPrompt,
                "messages", List.of(Map.of("role", "user", "content", userMessage))
            );

            String response = buildClient().post()
                    .uri("/v1/messages")
                    .bodyValue(body)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            JsonNode json = objectMapper.readTree(response);
            return json.path("content").get(0).path("text").asText();

        } catch (Exception e) {
            log.error("Erreur appel Claude: {}", e.getMessage());
            throw new RuntimeException("Claude API error: " + e.getMessage(), e);
        }
    }

    public <T> T completeAsJson(String systemPrompt, String userMessage, Class<T> type) {
        String raw = complete(systemPrompt, userMessage);
        try {
            // Extraire le JSON si enveloppé dans du markdown
            String json = extractJson(raw);
            return objectMapper.readValue(json, type);
        } catch (Exception e) {
            log.error("Erreur parsing JSON Claude: {}\nRéponse brute: {}", e.getMessage(), raw);
            throw new RuntimeException("Failed to parse Claude JSON response", e);
        }
    }

    private String extractJson(String text) {
        int start = text.indexOf('{');
        int end = text.lastIndexOf('}');
        if (start >= 0 && end > start) {
            return text.substring(start, end + 1);
        }
        // Essayer avec tableau JSON
        start = text.indexOf('[');
        end = text.lastIndexOf(']');
        if (start >= 0 && end > start) {
            return text.substring(start, end + 1);
        }
        return text;
    }
}

File: backend/src/main/kotlin/com/radar/prospection/claude/ClaudeService.kt
package com.radar.prospection.claude

import com.fasterxml.jackson.databind.ObjectMapper
import com.radar.prospection.config.RadarProperties
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient

@Service
class ClaudeService(
    private val properties: RadarProperties,
    private val webClientBuilder: WebClient.Builder,
    private val objectMapper: ObjectMapper
) {
    private val log = LoggerFactory.getLogger(javaClass)

    private fun buildClient(): WebClient = webClientBuilder
        .baseUrl(properties.claude.baseUrl)
        .defaultHeader("x-api-key", properties.claude.apiKey)
        .defaultHeader("anthropic-version", "2023-06-01")
        .defaultHeader("content-type", "application/json")
        .build()

    fun complete(systemPrompt: String, userMessage: String): String {
        try {
            val body = mapOf(
                "model" to properties.claude.model,
                "max_tokens" to 1024,
                "system" to systemPrompt,
                "messages" to listOf(mapOf("role" to "user", "content" to userMessage))
            )

            val response = buildClient().post()
                .uri("/v1/messages")
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String::class.java)
                .block()!!

            val json = objectMapper.readTree(response)
            return json.path("content").get(0).path("text").asText()

        } catch (e: Exception) {
            log.error("Erreur appel Claude: {}", e.message)
            throw RuntimeException("Claude API error: ${e.message}", e)
        }
    }

    fun <T> completeAsJson(systemPrompt: String, userMessage: String, type: Class<T>): T {
        val raw = complete(systemPrompt, userMessage)
        try {
            return objectMapper.readValue(extractJson(raw), type)
        } catch (e: Exception) {
            log.error("Erreur parsing JSON Claude: {}\nRéponse brute: {}", e.message, raw)
            throw RuntimeException("Failed to parse Claude JSON response", e)
        }
    }

    private fun extractJson(text: String): String {
        var start = text.indexOf('{')
        var end = text.lastIndexOf('}')
        if (start >= 0 && end > start) return text.substring(start, end + 1)
        start = text.indexOf('[')
        end = text.lastIndexOf(']')
        if (start >= 0 && end > start) return text.substring(start, end + 1)
        return text
    }
}
